### PR TITLE
Esdt nft optimization vm common

### PR DESCRIPTION
**Pull request by sasurobert**
Esdt nft optimization vm common

After activation token metadata is saved into the system account and not into user account. This makes the trie smaller. After activation and on tranfers, the data will be deleted from the user and kept only at the system account.

In case of cross shard SFT and metaESDT transfers, the second transfer cross shard of the same token will cost less as there is no need to put the MetaData information into the SCR.Data.

### CODE REVIEW

File: process/smartContract/builtInFunctions/factory.go [hunk 1]
```diff
@@ -13,44 +13,44 @@ import (
...
+	ESDTMultiTransferEnableEpoch uint32
+	ESDTTransferRoleEnableEpoch  uint32
+	GlobalMintBurnDisableEpoch   uint32
+	ESDTTransferMetaEnableEpoch  uint32
+	OptimizeNFTStoreEnableEpoch  uint32
 }
 
 // CreateBuiltInFunctionContainer creates a container that will hold all the available built in functions
-func CreateBuiltInFunctionContainer(args ArgsCreateBuiltInFunctionContainer) (vmcommon.BuiltInFunctionContainer, error) {
+func CreateBuiltInFunctionContainer(args ArgsCreateBuiltInFunctionContainer) (vmcommon.BuiltInFunctionContainer, vmcommon.SimpleESDTNFTStorageHandler, error) {
```
Comment (bogdan-rosianu) [inline process/smartContract/builtInFunctions/factory.go:31]: maybe add a todo to refactor this:
this factory should return the `bContainerFactory` and on caller side you should get the container and the nft storage handler

File: process/smartContract/hooks/blockChainHook.go [hunk 2]
```diff
@@ -485,19 +496,7 @@ func (bh *BlockChainHookImpl) GetESDTToken(address []byte, tokenID []byte, nonce
...
-	value, err := userAcc.AccountDataHandler().RetrieveValue(esdtTokenKey)
-	if err != nil {
-		return nil, err
-	}
-	if len(value) == 0 {
-		return esdtData, nil
-	}
-
-	err = bh.marshalizer.Unmarshal(esdtData, value)
+	esdtData, _, err = bh.nftStorageHandler.GetESDTNFTTokenOnDestination(userAcc, esdtTokenKey, nonce)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:499]: is this backwards compatible? if calling `GetESDTToken` before the enable epoch

File: process/smartContract/process.go [hunk 3]
```diff
@@ -1290,6 +1293,13 @@ func (sc *scProcessor) ProcessIfError(
 
 	sc.txFeeHandler.ProcessTransactionFee(consumedFee, big.NewInt(0), txHash)
 
+	if sc.flagOptimizeNFTStore.IsSet() {
+		err = sc.blockChainHook.SaveNFTMetaDataToSystemAccount(tx)
```
Comment (raduchis) [inline process/smartContract/process.go:1297]: is it ok to be called on ProcessIfError?

File: factory/blockProcessorCreator.go [hunk 4]
```diff
@@ -1036,19 +1040,19 @@ func (pcf *processComponentsFactory) createVMFactoryMeta(
...
+		MapDNSAddresses:              mapDNSAddresses,
+		Marshalizer:                  pcf.coreData.InternalMarshalizer(),
+		Accounts:                     accounts,
+		ShardCoordinator:             pcf.bootstrapComponents.ShardCoordinator(),
+		EpochNotifier:                pcf.epochNotifier,
+		ESDTMultiTransferEnableEpoch: pcf.epochConfig.EnableEpochs.ESDTMultiTransferEnableEpoch,
+		ESDTTransferRoleEnableEpoch:  pcf.epochConfig.EnableEpochs.ESDTTransferRoleEnableEpoch,
+		GlobalMintBurnDisableEpoch:   pcf.epochConfig.EnableEpochs.GlobalMintBurnDisableEpoch,
+		ESDTTransferMetaEnableEpoch:  pcf.epochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
+		OptimizeNFTStoreEnableEpoch:  pcf.epochConfig.EnableEpochs.OptimizeNFTStoreEnableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply factory/blockProcessorCreator.go:1055]: it was not used. it was not needed - only in esdt.go - in system SC

File: process/smartContract/hooks/blockChainHook.go (hunk 2 above)
Comment (sasurobert) [reply process/smartContract/hooks/blockChainHook.go:499]: yes. epoch enable logic is inside the new nftStorageHandler

File: factory/blockProcessorCreator.go [hunk 5]
```diff
@@ -1036,19 +1040,19 @@ func (pcf *processComponentsFactory) createVMFactoryMeta(
...
-		MapDNSAddresses:                      mapDNSAddresses,
-		Marshalizer:                          pcf.coreData.InternalMarshalizer(),
-		Accounts:                             accounts,
-		ShardCoordinator:                     pcf.bootstrapComponents.ShardCoordinator(),
-		EpochNotifier:                        pcf.epochNotifier,
-		ESDTMultiTransferEnableEpoch:         pcf.epochConfig.EnableEpochs.ESDTMultiTransferEnableEpoch,
-		ESDTTransferRoleEnableEpoch:          pcf.epochConfig.EnableEpochs.ESDTTransferRoleEnableEpoch,
-		GlobalMintBurnDisableEpoch:           pcf.epochConfig.EnableEpochs.GlobalMintBurnDisableEpoch,
-		ESDTTransferMetaEnableEpoch:          pcf.epochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
-		ESDTNFTCreateOnMultiShardEnableEpoch: pcf.epochConfig.EnableEpochs.ESDTNFTCreateOnMultiShardEnableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply factory/blockProcessorCreator.go:1051]: it is not used - it was deleted

File: process/smartContract/process.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/process.go:1297]: yes. It must be called here - as we have to save the NFT metadata to the system account if the cross shard NFT transfer failed for whatever reason. As we will not send this information the second time.

File: genesis/process/disabled/simpleNFTStorageHandler.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/data/esdt"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+// SimpleNFTStorageHandler implements the SimpleNFTStorageHandler interface but does nothing as it is disabled
+type SimpleNFTStorageHandler struct {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply genesis/process/disabled/simpleNFTStorageHandler.go:12]: done

File: process/smartContract/builtInFunctions/factory.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/factory.go:31]: renamed.

**Review by bogdan-rosianu — APPROVED** [review]

File: process/smartContract/process.go (hunk 3 above)
Comment (raduchis) [reply process/smartContract/process.go:1297]: ok!

File: genesis/process/disabled/simpleNFTStorageHandler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply genesis/process/disabled/simpleNFTStorageHandler.go]: ok

File: factory/blockProcessorCreator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/blockProcessorCreator.go:1051]: ok

**Review by raduchis — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.
image

**Pull request merged**